executor/src/main/java/com/redhat/service/smartevents/executor/ExecutorService.java

```diff
@@ -48,64 +60,127 @@ public class ExecutorService {
-
-            Map<String, Object> extensions = wrapHeadersToExtensionsMap(headers);
-            for (Map.Entry<String, Object> kv : extensions.entrySet()) {
-                // We currently support only String extensions
-                cloudEventBuilder.withExtension(kv.getKey(), kv.getValue().toString());
-            }
-            return cloudEventBuilder.build();
+            JsonNode payload = event == null ? mapper.createObjectNode() : mapper.readTree(event);
+
+            CloudEventData data = JsonCloudEventData.wrap(payload);
+
+            return wrapToCloudEvent(
+                    String.format("%sSource", executor.getProcessor().getDefinition().getRequestedSource().getType()),
+                    data,
+                    toExtensionsMap(headers));
         } catch (JsonProcessingException e2) {
             LOG.error("JsonProcessingException when generating CloudEvent for '{}'", event, e2);
             throw new CloudEventDeserializationException("Failed to generate event map");
         }
     }
 
-    public static Map<String, Object> wrapHeadersToExtensionsMap(Headers kafkaHeaders) {
-        Map<String, Object> resultMap = new HashMap<>();
-        if (kafkaHeaders == null) {
-            return resultMap;
+    private CloudEvent toErrorHandlerCloudEvent(String event) {
+        try {
+            // try to decode as cloud event
+            return CloudEventUtils.decode(event);
+        } catch (CloudEventDeserializationException e) {
+            // if it fails (e.g. for connector errors) try wrapping it
+            CloudEventData data = BytesCloudEventData.wrap(event.getBytes(StandardCharsets.UTF_8));
+            return wrapToCloudEvent("RhoseError", data, Collections.emptyMap());
         }
+    }
 
-        for (Header kh : kafkaHeaders) {
-            String cloudEventsExtensionName = adjustExtensionName(kh.key());
-            String headerValue = new String(kh.value(), StandardCharsets.UTF_8);
-            resultMap.put(cloudEventsExtensionName, headerValue);
+    private CloudEvent wrapToCloudEvent(String type, CloudEventData data, Map<String, String> extensions) {
+        CloudEventBuilder cloudEventBuilder = CloudEventBuilder.v1()
+                .withId(UUID.randomUUID().toString())
+                .withSource(URI.create(CLOUD_EVENT_SOURCE))
+                .withType(type)
+                .withData(data);
+
+        extensions.forEach(cloudEventBuilder::withExtension);
+
+        return cloudEventBuilder.build();
+    }
+
+    static String getOriginalEventId(CloudEvent cloudEvent, KafkaRecord<Integer, String> message) {
+        if (cloudEvent != null) {
+            return cloudEvent.getId();
+        }
+        if (message.getKey() != null) {
+            return message.getKey().toString();
+        }
+        if (message.getTimestamp() != null) {
+            return message.getTimestamp().toString();
+        }
+        return "unknown";
```

Comment:
If this is not supposed to happen, please add a log `error` too

Reply:
done :+1: